ghost/core/core/server/adapters/storage/S3Storage.ts

```diff
@@ -127,6 +157,106 @@ export default class S3Storage extends StorageBase {
         return `${this.cdnUrl}/${key}`;
     }
 
+    private async uploadMultipart(file: UploadFile, key: string, fileSize: number): Promise<string> {
+        let uploadId: string | undefined;
+
+        try {
+            // Step 1: Initiate multipart upload
+            const createResponse = await this.client.send(new CreateMultipartUploadCommand({
+                Bucket: this.bucket,
+                Key: key,
+                ContentType: file.type
+            }));
+
+            uploadId = createResponse.UploadId;
+            if (!uploadId) {
+                throw new errors.InternalServerError({
+                    message: tpl(messages.multipartUploadInitFailed)
+                });
+            }
+
+            // Step 2: Upload parts
+            const parts: {ETag: string; PartNumber: number}[] = [];
+            const fileHandle = await fs.promises.open(file.path, 'r');
+            
+            try {
+                let partNumber = 1;
+                let uploadedBytes = 0;
+
+                while (uploadedBytes < fileSize) {
+                    // Read one part into buffer
+                    const remainingBytes = fileSize - uploadedBytes;
+                    const currentPartSize = Math.min(this.partSize, remainingBytes);
+                    const buffer = Buffer.alloc(currentPartSize);
+                    
+                    const {bytesRead} = await fileHandle.read(buffer, 0, currentPartSize, uploadedBytes);
+                    
+                    if (bytesRead === 0) {
+                        throw new errors.InternalServerError({
+                            message: tpl(messages.multipartUploadReadFailed)
+                        });
+                    }
+
+                    // Upload this part
+                    const uploadPartResponse = await this.client.send(new UploadPartCommand({
+                        Bucket: this.bucket,
+                        Key: key,
+                        UploadId: uploadId,
+                        PartNumber: partNumber,
+                        Body: buffer.slice(0, bytesRead)
+                    }));
+
+                    if (!uploadPartResponse.ETag) {
+                        throw new errors.InternalServerError({
+                            message: tpl(messages.multipartUploadPartFailed, {partNumber})
+                        });
+                    }
+
+                    parts.push({
+                        ETag: uploadPartResponse.ETag,
+                        PartNumber: partNumber
+                    });
+
+                    uploadedBytes += bytesRead;
+                    partNumber += 1;
+
+                    const progress = Math.round((uploadedBytes / fileSize) * 100);
+                    logging.info(`Uploaded part ${partNumber - 1}/${Math.ceil(fileSize / this.partSize)} (${progress}%)`);
```

Comment:
I think moving this before the increment makes it easier to understand, I couldn't immediately see why we do `+ 1`